File: ios-native/Sprout/Model/Species.swift
import Foundation

/// Что увидела система: ярлык и уверенность. Свой тип, а не наблюдение
/// Vision, — чтобы разбор ярлыков проверялся без телефона.
struct Sighting: Hashable, Sendable {
    var name: String
    var confidence: Double
}

struct Guess: Hashable, Sendable {
    var species: String
    var dryingDays: Double
    var confidence: Double
}

/// Перевод ярлыков классификатора в вид растения. Это подсказка, а не
/// определитель: кактус от папоротника он отличит, замиокулькас от
/// сансевиерии — нет, поэтому оба поля правятся руками.
enum Species {
    /// Слова и сроки сушки — от частного к общему. Идём по списку, а не по
    /// уверенности: самый уверенный ярлык почти всегда самый общий, «plant».
    static let table: [(word: String, species: String, days: Double)] = [
        ("cactus", "Кактус", 30),
        ("succulent", "Суккулент", 21),
        ("aloe", "Алоэ", 14),
        ("jade", "Толстянка", 12),
        ("bonsai", "Бонсай", 4),
        ("orchid", "Орхидея", 9),
        ("fern", "Папоротник", 5),
        ("palm", "Пальма", 11),
        ("bamboo", "Бамбук", 6),
        ("ivy", "Плющ", 8),
        ("monstera", "Монстера", 9),
        ("philodendron", "Филодендрон", 8),
        ("ficus", "Фикус", 7),
        ("dracaena", "Драцена", 8),
        ("violet", "Фиалка", 4.5),
        ("tulip", "Тюльпан", 9),
        // Слово, внутри которого есть другое слово списка, стоит выше него:
        // «rosemary» содержит «rose».
        ("rosemary", "Розмарин", 5.9),
        ("rose", "Роза", 5),
        ("lily", "Лилия", 6),
        ("daisy", "Ромашка", 5),
        ("sunflower", "Подсолнух", 4),
        ("geranium", "Пеларгония", 5.4),
        ("begonia", "Бегония", 6.2),
        ("basil", "Базилик", 3.8),
        ("mint", "Мята", 4.2),
        ("herb", "Зелень", 4),
        ("moss", "Мох", 3),
        ("seedling", "Росток", 4),
        ("bush", "Кустик", 8),
        ("shrub", "Кустик", 8),
        ("tree", "Деревце", 12),
        ("grass", "Трава", 4),
        ("houseplant", "Комнатное растение", 7),
        ("flower", "Цветок", 5),
        ("leaf", "Комнатное растение", 7),
        ("plant", "Комнатное растение", 7),
    ]

    /// Порог низкий нарочно: уверенность делится на тысячи ярлыков, и даже
    /// верный редко берёт больше трети.
    static let floor = 0.06

    /// Пусто — ни одного знакомого слова; придумывать вид тогда нечестно.
    static func read(_ seen: [Sighting], floor: Double = Species.floor)
        -> Guess? {
        let strong = seen.filter { $0.confidence >= floor }
        for entry in table {
            // Наблюдения идут от самого уверенного, поэтому первое найденное
            // — лучшее.
            guard let hit = strong.first(where: {
                $0.name.lowercased().contains(entry.word)
            }) else { continue }
            return Guess(species: entry.species, dryingDays: entry.days,
                         confidence: hit.confidence)
        }
        return nil
    }

    /// Потолок барабана срока: дольше без воды не живёт и кактус.
    static let longest = 60

    /// Дробное — одним знаком и в родительном: «6,5 дня».
    static func periodLabel(_ days: Double) -> String {
        let whole = days.rounded()
        guard abs(days - whole) < 0.05 else {
            return Lang.format("Раз в %@ дня", Lang.decimal(days))
        }
        return Lang.format("Раз в %lld дней", Int(whole))
    }

    /// Срок в середине фразы: «раз в 6 дней».
    static func periodPhrase(_ days: Double) -> String {
        let whole = days.rounded()
        guard abs(days - whole) < 0.05 else {
            return Lang.format("раз в %@ дня", Lang.decimal(days))
        }
        return Lang.format("раз в %lld дней", Int(whole))
    }

    /// Обычный срок для вида, вписанного по-русски, — из той же таблицы.
    static func usual(for species: String) -> Double? {
        let name = species.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return nil }
        return table.first {
            $0.species.compare(name, options: .caseInsensitive) == .orderedSame
        }?.days
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/model/Species.kt
package com.ledro6.sprout.model

import kotlin.math.abs

/** Что увидела система: ярлык и уверенность. */
data class Sighting(val name: String, val confidence: Double)

data class Guess(val species: String, val dryingDays: Double, val confidence: Double)

/**
 * Перевод ярлыков классификатора в вид растения. Это подсказка, а не
 * определитель: кактус от папоротника он отличит, замиокулькас от
 * сансевиерии — нет, поэтому оба поля правятся руками.
 */
object Species {
    data class Row(val word: String, val species: String, val days: Double)

    /** Слова и сроки — от частного к общему, как на iOS. */
    val table: List<Row> = listOf(
        Row("cactus", Lang.key("Кактус"), 30.0),
        Row("succulent", Lang.key("Суккулент"), 21.0),
        Row("aloe", Lang.key("Алоэ"), 14.0),
        Row("jade", Lang.key("Толстянка"), 12.0),
        Row("bonsai", Lang.key("Бонсай"), 4.0),
        Row("orchid", Lang.key("Орхидея"), 9.0),
        Row("fern", Lang.key("Папоротник"), 5.0),
        Row("palm", Lang.key("Пальма"), 11.0),
        Row("bamboo", Lang.key("Бамбук"), 6.0),
        Row("ivy", Lang.key("Плющ"), 8.0),
        Row("monstera", Lang.key("Монстера"), 9.0),
        Row("philodendron", Lang.key("Филодендрон"), 8.0),
        Row("ficus", Lang.key("Фикус"), 7.0),
        Row("dracaena", Lang.key("Драцена"), 8.0),
        Row("violet", Lang.key("Фиалка"), 4.5),
        Row("tulip", Lang.key("Тюльпан"), 9.0),
        // «rosemary» содержит «rose» — поэтому выше.
        Row("rosemary", Lang.key("Розмарин"), 5.9),
        Row("rose", Lang.key("Роза"), 5.0),
        Row("lily", Lang.key("Лилия"), 6.0),
        Row("daisy", Lang.key("Ромашка"), 5.0),
        Row("sunflower", Lang.key("Подсолнух"), 4.0),
        Row("geranium", Lang.key("Пеларгония"), 5.4),
        Row("begonia", Lang.key("Бегония"), 6.2),
        Row("basil", Lang.key("Базилик"), 3.8),
        Row("mint", Lang.key("Мята"), 4.2),
        Row("herb", Lang.key("Зелень"), 4.0),
        Row("moss", Lang.key("Мох"), 3.0),
        Row("seedling", Lang.key("Росток"), 4.0),
        Row("bush", Lang.key("Кустик"), 8.0),
        Row("shrub", Lang.key("Кустик"), 8.0),
        Row("tree", Lang.key("Деревце"), 12.0),
        Row("grass", Lang.key("Трава"), 4.0),
        Row("houseplant", Lang.key("Комнатное растение"), 7.0),
        Row("flower", Lang.key("Цветок"), 5.0),
        Row("leaf", Lang.key("Комнатное растение"), 7.0),
        Row("plant", Lang.key("Комнатное растение"), 7.0),
    )

    /** Порог низкий нарочно: уверенность делится на сотни ярлыков. */
    const val FLOOR = 0.06

    /** Пусто — ни одного знакомого слова; придумывать вид тогда нечестно. */
    fun read(seen: List<Sighting>, floor: Double = FLOOR): Guess? {
        val strong = seen.filter { it.confidence >= floor }
            .sortedByDescending { it.confidence }
        for (entry in table) {
            val hit = strong.firstOrNull { it.name.lowercase().contains(entry.word) } ?: continue
            return Guess(Lang.text(entry.species), entry.days, hit.confidence)
        }
        return null
    }

    /** Потолок барабана срока: дольше без воды не живёт и кактус. */
    const val LONGEST = 60

    /** Дробное — одним знаком и в родительном: «6,5 дня». */
    fun periodLabel(days: Double): String {
        val whole = days.rounded()
        if (abs(days - whole) >= 0.05) return Lang.format("Раз в %@ дня", Lang.decimal(days))
        return Lang.format("Раз в %lld дней", whole.toInt())
    }

    fun periodPhrase(days: Double): String {
        val whole = days.rounded()
        if (abs(days - whole) >= 0.05) return Lang.format("раз в %@ дня", Lang.decimal(days))
        return Lang.format("раз в %lld дней", whole.toInt())
    }

    /** Обычный срок для вида — на языке телефона или по-русски. */
    fun usual(species: String): Double? {
        val name = species.trim()
        if (name.isEmpty()) return null
        return table.firstOrNull { row ->
            Lang.text(row.species).equals(name, ignoreCase = true) ||
                row.species.equals(name, ignoreCase = true)
        }?.days
    }
}
